vmms/distDocker.py

```diff
@@ -0,0 +1,315 @@
+#
+# distDocker.py
+#
+# Implements the Tango VMMS interface to run Tango jobs in 
+# docker containers on a list of host machines. This list of
+# host machines must be able to run docker and be accessible
+# by SSH. The IP address of the host machine is stored in the
+# `domain_name` attribtue of TangoMachine.
+#
+
+import random, subprocess, re, time, logging, threading, os, sys, shutil
+import config
+from tangoObjects import TangoMachine
+
+def timeout(command, time_out=1):
+    """ timeout - Run a unix command with a timeout. Return -1 on
+    timeout, otherwise return the return value from the command, which
+    is typically 0 for success, 1-255 for failure.
+    """ 
+
+    # Launch the command
+    p = subprocess.Popen(command,
+                        stdout=open("/dev/null", 'w'),
+                        stderr=subprocess.STDOUT)
+
+    # Wait for the command to complete
+    t = 0.0
+    while t < time_out and p.poll() is None:
+        time.sleep(config.Config.TIMER_POLL_INTERVAL)
+        t += config.Config.TIMER_POLL_INTERVAL
+
+    # Determine why the while loop terminated
+    if p.poll() is None:
+        subprocess.call(["/bin/kill", "-9", str(p.pid)])
+        returncode = -1
+    else:
+        returncode = p.poll()
+    return returncode
+
+def timeoutWithReturnStatus(command, time_out, returnValue = 0):
+    """ timeoutWithReturnStatus - Run a Unix command with a timeout,
+    until the expected value is returned by the command; On timeout,
+    return last error code obtained from the command.
+    """
+    p = subprocess.Popen(command, 
+                        stdout=open("/dev/null", 'w'), 
+                        stderr=subprocess.STDOUT)
+    t = 0.0
+    while (t < time_out):
+        ret = p.poll()
+        if ret is None:
+            time.sleep(config.Config.TIMER_POLL_INTERVAL)
+            t += config.Config.TIMER_POLL_INTERVAL
+        elif ret == returnValue:
+            return ret
+        else:
+            p = subprocess.Popen(command,
+                            stdout=open("/dev/null", 'w'),
+                            stderr=subprocess.STDOUT)
+    return ret
+
+#
+# User defined exceptions
+#
+
+class DistDocker:
+
+    _SSH_FLAGS = ["-q", "-i", "/Users/Mihir/Documents/prog/Autolab/mp_tango.pem",
+              "-o", "StrictHostKeyChecking=no",
+              "-o", "GSSAPIAuthentication=no"]
+
+    def __init__(self):
+        """ Checks if the machine is ready to run docker containers.
+        Initialize boot2docker if running on OS X.
+        """
+        try:
+            self.log = logging.getLogger("DistDocker")
+            self.hosts = ['54.186.238.205', '54.68.89.235']
```

Comment:
I'm sure you know this, but this will want to be factored out into a config parameter eventually.